File: src/main/java/com/pubnub/api/endpoints/message_actions/RemoveMessageAction.java
package com.pubnub.api.endpoints.message_actions;

import com.pubnub.api.PubNub;
import com.pubnub.api.PubNubException;
import com.pubnub.api.builder.PubNubErrorBuilder;
import com.pubnub.api.endpoints.Endpoint;
import com.pubnub.api.enums.PNOperationType;
import com.pubnub.api.managers.RetrofitManager;
import com.pubnub.api.managers.TelemetryManager;
import com.pubnub.api.managers.token_manager.TokenManager;
import com.pubnub.api.models.consumer.message_actions.PNRemoveMessageActionResult;
import lombok.Setter;
import lombok.experimental.Accessors;
import retrofit2.Call;
import retrofit2.Response;

import java.util.Collections;
import java.util.List;
import java.util.Map;

@Accessors(chain = true, fluent = true)
public class RemoveMessageAction extends Endpoint<Object, PNRemoveMessageActionResult> {

    @Setter
    private String channel;

    @Setter
    private Long messageTimetoken;

    @Setter
    private Long actionTimetoken;

    public RemoveMessageAction(PubNub pubnubInstance,
                               TelemetryManager telemetry,
                               RetrofitManager retrofitInstance,
                               TokenManager tokenManager) {
        super(pubnubInstance, telemetry, retrofitInstance, tokenManager);
    }

    @Override
    protected List<String> getAffectedChannels() {
        return Collections.singletonList(channel);
    }

    @Override
    protected List<String> getAffectedChannelGroups() {
        return null;
    }

    @Override
    protected void validateParams() throws PubNubException {
        if (channel == null || channel.isEmpty()) {
            throw PubNubException.builder().pubnubError(PubNubErrorBuilder.PNERROBJ_CHANNEL_MISSING).build();
        }
        if (messageTimetoken == null) {
            throw PubNubException.builder().pubnubError(PubNubErrorBuilder.PNERROBJ_MESSAGE_TIMETOKEN_MISSING).build();
        }
        if (actionTimetoken == null) {
            throw PubNubException.builder().pubnubError(PubNubErrorBuilder.PNERROBJ_MESSAGE_ACTION_TIMETOKEN_MISSING)
                    .build();
        }
        if (this.getPubnub().getConfiguration().getSubscribeKey() == null || this.getPubnub()
                .getConfiguration()
                .getSubscribeKey()
                .isEmpty()) {
            throw PubNubException.builder().pubnubError(PubNubErrorBuilder.PNERROBJ_SUBSCRIBE_KEY_MISSING).build();
        }
    }

    @Override
    protected Call<Object> doWork(Map<String, String> params) throws PubNubException {

        params.putAll(encodeParams(params));

        return this.getRetrofit()
                .getMessageActionService()
                .deleteMessageAction(this.getPubnub().getConfiguration().getSubscribeKey(), channel,
                        Long.toString(messageTimetoken).toLowerCase(), Long.toString(actionTimetoken).toLowerCase(),
                        params);
    }

    @Override
    protected PNRemoveMessageActionResult createResponse(Response<Object> input) throws PubNubException {
        return new PNRemoveMessageActionResult();
    }

    @Override
    protected PNOperationType getOperationType() {
        return PNOperationType.PNDeleteMessageAction;
    }

    @Override
    protected boolean isAuthRequired() {
        return true;
    }
}

File: src/main/kotlin/com/pubnub/api/endpoints/message_actions/RemoveMessageAction.kt
package com.pubnub.api.endpoints.message_actions

import com.pubnub.api.Endpoint
import com.pubnub.api.PubNub
import com.pubnub.api.PubNubError
import com.pubnub.api.PubNubException
import com.pubnub.api.enums.PNOperationType
import com.pubnub.api.models.consumer.message_actions.PNRemoveMessageActionResult
import com.pubnub.api.retry.RetryableEndpointGroup
import retrofit2.Call
import retrofit2.Response
import java.util.Locale

/**
 * @see [PubNub.removeMessageAction]
 */
class RemoveMessageAction internal constructor(
    pubnub: PubNub,
    val channel: String,
    val messageTimetoken: Long,
    val actionTimetoken: Long
) : Endpoint<Void, PNRemoveMessageActionResult>(pubnub) {

    override fun validateParams() {
        super.validateParams()
        if (channel.isBlank()) throw PubNubException(PubNubError.CHANNEL_MISSING)
    }

    override fun getAffectedChannels() = listOf(channel)

    override fun doWork(queryParams: HashMap<String, String>): Call<Void> {
        return pubnub.retrofitManager.messageActionService
            .deleteMessageAction(
                subKey = pubnub.configuration.subscribeKey,
                channel = channel,
                messageTimetoken = messageTimetoken.toString().lowercase(Locale.getDefault()),
                actionTimetoken = actionTimetoken.toString().lowercase(Locale.getDefault()),
                options = queryParams
            )
    }

    override fun createResponse(input: Response<Void>): PNRemoveMessageActionResult? {
        return PNRemoveMessageActionResult()
    }

    override fun operationType() = PNOperationType.PNDeleteMessageAction

    override fun getEndpointGroupName(): RetryableEndpointGroup = RetryableEndpointGroup.MESSAGE_REACTION
}
